pkg/connection/store/kubernetes/store.go

```diff
@@ -0,0 +1,168 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubernetes
+
+import (
+	"context"
+	"encoding/json"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/crossplane/crossplane-runtime/pkg/connection/store"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+)
+
+// Error strings.
+const (
+	errGetSecret            = "cannot get secret"
+	errDeleteSecret         = "cannot delete secret"
+	errUpdateSecret         = "cannot update secret"
+	errCreateOrUpdateSecret = "cannot create or update connection applicator"
+	errParseMetadata        = "cannot parse metadata"
+
+	errExtractKubernetesAuthCreds = "cannot extract kubernetes auth credentials"
+)
+
+type secretMetadata struct {
+	Labels      map[string]string `json:"labels"`
+	Annotations map[string]string `json:"annotations"`
+	Type        corev1.SecretType `json:"type"`
+}
+
+// SecretStore is a Kubernetes Secret Store.
+type SecretStore struct {
+	client     client.Client
+	applicator resource.Applicator
+
+	defaultNamespace string
+}
+
+// NewSecretStore returns a new Kubernetes SecretStore.
+func NewSecretStore(ctx context.Context, local client.Client, cfg v1.SecretStoreConfig) (*SecretStore, error) {
+	if cfg.Kubernetes == nil {
+		// No KubernetesSecretStoreConfig provided, local API Server will be
+		// used as Secret Store.
+		return &SecretStore{
+			client: local,
+			applicator: resource.NewApplicatorWithRetry(resource.NewAPIPatchingApplicator(local),
+				resource.IsAPIErrorWrapped, nil),
+			defaultNamespace: cfg.DefaultScope,
+		}, nil
+	}
+
+	// Configure client for an external API server with a given Kubeconfig.
+	kfg, err := resource.CommonCredentialExtractor(ctx, cfg.Kubernetes.Auth.Source, local, cfg.Kubernetes.Auth.CommonCredentialSelectors)
+	if err != nil {
+		return nil, errors.Wrap(err, errExtractKubernetesAuthCreds)
+	}
+	remote, err := clientForKubeconfig(kfg)
+	if err != nil {
+		return nil, errors.Wrap(err, errExtractKubernetesAuthCreds)
+	}
+
+	return &SecretStore{
+		client: remote,
+		applicator: resource.NewApplicatorWithRetry(resource.NewAPIPatchingApplicator(remote),
+			resource.IsAPIErrorWrapped, nil),
+		defaultNamespace: cfg.DefaultScope,
+	}, nil
+}
+
+// ReadKeyValues reads and returns key value pairs for a given Kubernetes Secret.
+func (ss *SecretStore) ReadKeyValues(ctx context.Context, i store.Secret) (store.KeyValues, error) {
+	s := &corev1.Secret{}
+	return s.Data, errors.Wrapf(ss.client.Get(ctx, types.NamespacedName{Name: i.Name, Namespace: ss.namespaceForSecret(i)}, s), errGetSecret)
```

Comment:
Nit: Wrapf is only necessary when there's string formatting in the error string.

```suggestion
	return s.Data, errors.Wrap(ss.client.Get(ctx, types.NamespacedName{Name: i.Name, Namespace: ss.namespaceForSecret(i)}, s), errGetSecret)
```

Reply:
Yeah, was not intentional :)